### PR TITLE
Warn when an escaped newline skips multiple lines

### LINKED ISSUE #87319: Warn about strings that skip multiple newlines with one backslash
(This came up in today's @rust-lang/lang meeting.)

Given the following code:

```rust
fn main() {
    let s1 = "a\

b";
    let s2 = "a\

    b";
    println!("{:?} {:/}", s1, s2);
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a783aa21f2ec4064d0ae0cab0f4c8777)

The compiler interprets both of the string literals as `"ab"`, omitting multiple newlines after the `\`.

This seems confusing, and different from the behavior of other languages. For instance, C and Python both produce parse errors for such string literals. Or, for a Python multi-line string literal, the backslash escapes the immediately following newline but not the subsequent newline.

Rust does allow unescaped newlines in string literals, which is *useful*. The ability to escape a newline can also be useful sometimes. But the combination of both escaped and unescaped newlines seems sufficiently confusing and surprising to merit a warning.

We'd ideally like to work towards this being an error. Depending on the presence of existing usage in the ecosystem, that might or might not require an edition.

### PR DESCRIPTION
**Pull request by jesyspa**
Resolves #87319

### CODE REVIEW

**Comment by jesyspa** [comment]
Hi @estebank ,
Sorry for the ping, I was wondering if you could review this since I'm guessing @cjgillot is away?  It's very similar to my [PR 87596](https://github.com/rust-lang/rust/pull/87596) that you reviewed earlier.

**Review by estebank — COMMENTED** [review]
Changes look reasonable, left some nitpicks.

File: compiler/rustc_lexer/src/unescape.rs [hunk 1]
```diff
@@ -320,6 +324,11 @@ where
             .bytes()
             .position(|b| b != b' ' && b != b'\t' && b != b'\n' && b != b'\r')
             .unwrap_or(str.len());
+        if str[1..first_non_space].contains('\n') {
```
[the remark this answers is not in the document]
Comment (jesyspa) [reply compiler/rustc_lexer/src/unescape.rs:327]: Makes sense.  The two names that come to mind for this are `as_str` (but might be confusing as it sounds like a method) or `tail` (which then gets shadowed when we're done).  I'm leaning towards `tail`, what do you think? (`s` is also possible given the small scope, but it seems a bit minimalistic.)

[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_lexer/src/unescape.rs:327]: `tail` works.

File: src/test/ui/fmt/format-string-error-2.rs [hunk 2]
```diff
@@ -5,7 +5,7 @@ fn main() {
     a");
     //~^ ERROR invalid format string
     format!("{ \
-
+               \
```
[the remark this answers is not in the document]
Comment (jesyspa) [reply src/test/ui/fmt/format-string-error-2.rs:8]: Sounds good!  I've added one, plus another for the last PR.

File: compiler/rustc_lexer/src/unescape.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jesyspa) [reply compiler/rustc_lexer/src/unescape.rs]: Done!  Thanks for the review!

**Pull request merged**